<a href="https://colab.research.google.com/github/theweird-kid/ML-Notes/blob/main/Preprocessing/Handling_imbalanced_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Imbalanced Dataset:**
 dataset with an unequal class Distribution

In [1]:
import pandas as pd
import numpy as np

In [2]:
# Load dataset
credit_df = pd.read_csv('credit_data.csv')

In [3]:
credit_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
credit_df.shape

(284807, 31)

In [5]:
# distribution of the two classes
credit_df['Class'].value_counts()

Class
0    284315
1       492
Name: count, dtype: int64

This is highly Imabalanced Dataset
* 0 -> Legit Transactions
* 1 -> Fraudlent Transactions

In [6]:
# Separating the legit and Fradulent Transactions
legit = credit_df[credit_df.Class == 0]
fraud = credit_df[credit_df.Class == 1]

In [7]:
print(legit.shape, fraud.shape)

(284315, 31) (492, 31)


Undersampling to Handle Imbalanced Data

Building a sampled dataset containing similar Distribution of legit and Fraudlent Transactions

Number of Fraudlent Tx -> 492
Number of Legit Tx -> 284315

In [8]:
legit_sample = legit.sample(n=492)

Concatenate the 2 DataFrames

In [10]:
new_dataset = pd.concat([legit_sample, fraud], axis=0)

In [11]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
173237  121399.0 -1.934082  0.477274  0.384825 -2.272628 -0.905010 -0.954896   
111565   72276.0  0.944020 -0.895521  0.892340 -0.178583 -1.282552 -0.234313   
42170    40979.0 -2.317544  0.139915  1.667695  0.996202 -1.319787  1.562604   
34686    37782.0  1.190475 -1.774744  1.413539 -1.317311 -2.076222  1.030123   
242374  151469.0 -1.088843  1.351614  1.685797  1.118713 -0.119331  0.161590   

              V7        V8        V9  ...       V21       V22       V23  \
173237  0.003317  0.788140  0.811608  ...  0.274167  0.741763 -0.181849   
111565 -0.653505  0.009005  0.851177  ...  0.256035  0.412140 -0.209247   
42170   0.227558  0.812674  0.102422  ...  0.036372  0.509297  0.086616   
34686  -2.093246  0.546992 -1.079955  ...  0.168181  0.895351 -0.020314   
242374  0.408187  0.496664 -0.386055  ... -0.484375 -1.142462 -0.086927   

             V24       V25       V26       V27       V28  Amount  Class  
173237  0.077416 -0.051461  0.530904  0.219801  0.051511   90.00      0  
111565 -0.017456  0.009069  1.506750 -0.087412  0.042023  179.40      0  
42170  -0.299821  0.020668 -0.291283 -0.600431 -0.699879  182.08      0  
34686  -0.277083  0.034098  0.031385  0.108489  0.029906   74.50      0  
242374 -0.041445  0.464273 -0.784378  0.241238  0.076213   25.71      0  

[5 rows x 31 columns]

In [12]:
new_dataset['Class'].value_counts()

Class
0    492
1    492
Name: count, dtype: int64